### ISSUE TITLE
Scheduled Date not working

### ISSUE THREAD

**Issue by lordscarlet** [nathanwoulfe/Plumber-2#117]
**Describe the bug**
I can't imagine this has been around this long and no one else has run into it, but...

A publish request with a scheduled date does not work. I have tried a few different scenarios, and I can't seem to get any work. I concede this could certainly be a configuration or our specific implementation issue, but.. I have tried it in multiple environments, and it just does not seem to work for us.

**To Reproduce**
1. Go to a node
2. Make a change
3. Request publish
4. Specify a scheduled date

**Expected behavior**
If approval happens before the scheduled date and time, when that timestamp arrives, the content will be available as the published page

**Plumber version**
`2.02`

**Licensed version?**
Yes

**Additional context**
I don't know if Plumber relies on it, but I tested a Scheduled Publish without going through the Plumber workflow (using an admin), and it functioned as expected.

I don't begin to be an expert in the Umbraco database, or particularly the workflow tables, but the `SourceVersion` does not seem to match up to the version id in either `umbracoContentVersion` or `umbracoDocumentVersion`. I have no idea if that means anything.

Additionally, I am trying to cobble together a query to determine nodes that:

1. Were scheduled to publish
2. Were approved
3. The version that was scheduled to publish did not actually get published

I haven't quite gotten there yet, but I think I am close. This is an issue in our production instance, so if we can do quick queries and manually publish anything that has an issue, that would be beneficial.

**Comment by nathanwoulfe** [comment]
Thanks @lordscarlet, will have a dig into it...

I know I changed how the scheduling worked way back in the day, so that the current implementation uses the Umbraco UI elements to work the same way as scheduling a non-workflowed publish. In theory, it should just work because the scheduling component is all native Umbraco stuff. 

I have a feeling it might be related to setting a scheduled date on a node with unsaved changes, then submitting for publish. Somehow the schedule isn't being set or updated. 

Not sure, like I said needs some investigation. Anything else you might discover would be super helpful too.

**Comment by lordscarlet** [comment]
I am pretty certain I specifically hit save before requesting publish, but
I can't say with 100% certainty. I am on the other side of the world from
you, so I am off to bed, but will test some of this out in the morning.

Do you know what is saved in the umbraco database if something is
scheduled? I can take a look in there to see how those records compare to
ones triggered by plumber.

**Comment by lordscarlet** [comment]
I BELIEVE that instance is working (approving beyond the published date)
but, again, can't say for sure at this moment.

The first thing to check is probably just if you can replicate the problem
and it's not something off in my setup.

**Comment by nathanwoulfe** [comment]
I agree, I think that case is ok - handler runs on the publishing event, checks for any active workflow. If it finds one, the publish is cancelled and should clear the scheduled date (since that date must have passed to trigger the publishing event, given the node can't be published from the UI while it's in a workflow). 

Wondering if that's somehow interacting with the other scheduling flow...

V7 Plumber didn't work with scheduling, this is probably why!

**Comment by lordscarlet** [comment]
I apparently had an update to my original submission that I never saved. It was:

The real problem seems to be that there is no `DestVersion`? On a scheduled publish, should that be there? Or does it not populate until it has actually published?

**Comment by lordscarlet** [comment]
It looks like when I do a scheduled publish via the Umbraco UI, it adds a record to `umbracoContentSchedule`. When I do the same through the Plumber interface, no such record is added.

**Comment by nathanwoulfe** [comment]
That sounds pretty conclusive to me! 

The source and dest versions in the plumber table are for comparing differences for workflowed changes, dest should be populated when the workflow starts, will check that too. I don't think that's a super useful feature, given the diff is horrible for anything other than simple text fields. Preview is more helpful, imo.

**Comment by lordscarlet** [comment]
Related: we often get an error when trying to run the diff.

Also a reminder that we are on 2.0.2, not sure if there are any known
related changes in the intervening time.

**Comment by lordscarlet** [comment]
Let me know if there is anything I can do to debug.. I am going to keep digging on my end.

**Comment by lordscarlet** [comment]
> 

on a side note, I really want to take the time to write a better diff plugin. I haven't searched to see if anyone else has. But what I really want is more of a preview diff.

**Comment by lordscarlet** [comment]
I apparently had a vanilla install sitting around with `v2.0.1` of Plumber. I tested it, it worked.

I installed `v2.0.2`.. replicated the bug.

So this install does not contain any of our surface controllers or any of that, it's an out-of-the-box `8.17.0` install with Plumber installed.

I then updated to `v2.1.1` and I am seeing the same (bad) behavior. 

If there's any other information we can provide, please let me know.

**Comment by nathanwoulfe** [comment]
@lordscarlet have a look at 2.1.2-rc (on Nuget now), I think it resolves both the scheduling and content diff issues.

Scheduling wasn't being written to the Umbraco content schedule (as you discovered), diff wasn't setting the destVersion when creating a new workflow. I've done a quick test of both, but would appreciate another set of eyes.

**Comment by lordscarlet** [comment]
I have loaded this on my vanilla install, and things look good. Loading it on our real instance takes a little more work, particularly in verifying accessibility issues (although I think you implemented most/all of the changes submitted by @azdyrski).

I also will need that instance to test the diff, but it's looking good so far!
![image](https://user-images.githubusercontent.com/17951/151279286-f7ae09db-97f9-4c68-9f7e-918f76f17e11.png)

**Comment by nathanwoulfe** [comment]
Have you tried with a test license on a local site? Add test.lic to the plumber app plugins folder and you should get all features (provided the site is running in debug mode, from Visual Studio)

A11y stuff still needs a bit of work, tricky to separate the things in my control from the stuff I inherit fr Umbraco. 

Good to hear it's looking positive though 👍

**Comment by lordscarlet** [comment]
Our initial tests in our actual destined for production instances are successful as well.

Diff looks good as well.

![image](https://user-images.githubusercontent.com/17951/151442371-137353e6-f5ac-4311-9bef-531dbc879774.png)

**Comment by nathanwoulfe** [comment]
Awesome. Surprised it took this long for anyone to notice it wasn't working properly 😅

I'll wrap it up into a patch release today. 

Unrelated - have noticed in our prod environment random errors on some workflows, reported as PublishFailedInvalidContent or something similar. Interested if you've seen the same?

**Comment by lordscarlet** [comment]
@nathanwoulfe where does that error manifest itself? Is it handled by Umbraco and and in the log there, or are you seeing it in external monitoring of some sort?

**Comment by lordscarlet** [comment]
The only thing I see in the last week in Umbraco's log is below. I _think_ that is related to the diff bug you solved above?

I don't see anything in Application Insights.

```
System.InvalidOperationException: Sequence contains more than one element
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
   at NPoco.Database.SingleOrDefault[T](String sql, Object[] args)
   at Plumber.Core.Repositories.Implement.Database.Single[T](String query, Object[] args)
   at Plumber.Core.Repositories.Implement.ContentRepository.GetDestinationVersion(Int32 nodeId, Nullable`1 date)
   at Plumber.Core.Services.Implement.InstancesService.GetDiff(Guid guid)
   at Plumber.Web.Api.InstancesController.GetDiff(Guid guid)
```

**Comment by lordscarlet** [comment]
I do see some of these:

```
Plumber.Core.Exceptions.WorkflowException: Workflow instance 6937 is not pending any action.
   at Plumber.Core.Processes.WorkflowProcess.ActionWorkflow(WorkflowInstancePoco instance, WorkflowAction action, Int32 userId, String comment)
   at Plumber.Web.Api.ActionsController.Approve(TaskData model)
````

**Comment by nathanwoulfe** [comment]
Diff and schedule bugs were fixed in v2.1.2 👍
